src/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -130,11 +136,162 @@ QString DBusService::getLogs() {
+     * so that's conditional.
+     */
+    if (rv == EAI_NONAME || rv == EAI_FAIL ||
+#ifdef EAI_NODATA
+        rv == EAI_NODATA ||
+#endif
+        (retries >= 0 && !retries--)) {
+      logger.log() << "Failed to resolve the address endpoint";
+      return false;
+    }
+
+    logger.log() << "Trying again in" << (timeout / 1000000.0) << "seconds";
+    usleep(timeout);
+  }
+
+  if ((resolved->ai_family == AF_INET &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in)) ||
+      (resolved->ai_family == AF_INET6 &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in6))) {
+    memcpy(endpoint, resolved->ai_addr, resolved->ai_addrlen);
+    freeaddrinfo(resolved);
+    return true;
+  }
+
+  logger.log() << "Invalid endpoint" << address;
+  freeaddrinfo(resolved);
+  return false;
+}
+
+bool DBusService::switchServer(const Config& config) {
+  logger.log() << "Switching server";
+
+  wg_device* device = static_cast<wg_device*>(calloc(1, sizeof(*device)));
+  if (!device) {
+    logger.log() << "Allocation failure";
+    return false;
+  }
+
+  auto guard = qScopeGuard([&] { wg_free_device(device); });
+
+  strncpy(device->name, WG_INTERFACE, IFNAMSIZ - 1);
+  device->name[IFNAMSIZ - 1] = '\0';
+
+  wg_peer* peer = static_cast<wg_peer*>(calloc(1, sizeof(*peer)));
+  if (!peer) {
+    logger.log() << "Allocation failure";
+    return false;
+  }
+
+  peer->flags =
+      (wg_peer_flags)(WGPEER_HAS_PUBLIC_KEY | WGPEER_REPLACE_ALLOWEDIPS);
+
+  device->first_peer = device->last_peer = peer;
+  device->flags = WGDEVICE_REPLACE_PEERS;
+
+  if (!parseEndpoint(&peer->endpoint.addr, config.m_serverIpv4AddrIn,
+                     config.m_serverPort)) {
+    return false;
+  }
+
+  wg_key_from_base64(peer->public_key, config.m_serverPublicKey.toLocal8Bit());
+
+  for (const IPAddressRange& ip : config.m_allowedIPAddressRanges) {
+    wg_allowedip* allowedip =
+        static_cast<wg_allowedip*>(calloc(1, sizeof(*allowedip)));
+    if (!allowedip) {
+      logger.log() << "Allocation failure";
+      return false;
+    }
+
+    if (!peer->first_allowedip) {
+      peer->first_allowedip = allowedip;
+    } else {
+      peer->last_allowedip->next_allowedip = allowedip;
+    }
+
+    peer->last_allowedip = allowedip;
+    allowedip->cidr = ip.range();
+
+    bool ok = false;
+    if (ip.type() == IPAddressRange::IPv4) {
+      allowedip->family = AF_INET;
+      ok = inet_pton(AF_INET, ip.ipAddress().toLocal8Bit(), &allowedip->ip4) ==
+           1;
+    } else if (ip.type() == IPAddressRange::IPv6) {
+      allowedip->family = AF_INET6;
+      ok = inet_pton(AF_INET6, ip.ipAddress().toLocal8Bit(), &allowedip->ip6) ==
+           1;
+    } else {
+      logger.log() << "Invalid IPAddressRange type";
+      return false;
+    }
+
+    if (!ok) {
+      logger.log() << "Invalid IP address:" << ip.ipAddress();
+      return false;
+    }
+  }
+
+  if (wg_set_device(device) != 0) {
```

Comment:
just checking my understanding. after updating the peer ips, we are now setting device we have created using `wg_set_device`. 

i'm a little surprised this works, so just to confirm my understanding.

you allocated a new device, you gave it the name of our existing device, you set a new set of peers, but no other config. then set device just updated the peers because we had allocated the correct name and it could see it exists already.

Reply:
This is correct. We are updating the 'settings' of an existing device. The way WG does this is to call `wg_set_device` with a "partially-completed" device data struct. The flags are used to say which part of the device struct has been completed.

If you have time, take a look at the `wg set` command implementation:
1. https://github.com/WireGuard/wireguard-tools/blob/master/src/set.c#L25-L29
2. https://github.com/WireGuard/wireguard-tools/blob/master/src/config.c